packages/toolkit/src/createSlice.ts

```diff
@@ -673,30 +1089,51 @@ export function buildCreateSlice({ creators }: BuildCreateSliceConfig = {}) {
         context.sliceCaseReducersByName[name] = reducer
         return contextMethods
       },
+      getInitialState,
     }
 
-    reducerNames.forEach((reducerName) => {
-      const reducerDefinition = reducers[reducerName]
-      const reducerDetails: ReducerDetails = {
-        reducerName,
-        type: getType(name, reducerName),
-        createNotation: typeof options.reducers === 'function',
+    if (isCreatorCallback(options.reducers)) {
+      const reducers = options.reducers(creators as any)
+      for (const [reducerName, reducerDefinition] of Object.entries(reducers)) {
+        const { _reducerDefinitionType: type } = reducerDefinition
+        if (typeof type === 'undefined') {
+          throw new Error(
+            'Please use reducer creators passed to callback. Each reducer definition must have a `_reducerDefinitionType` property indicating which handler to use.',
+          )
+        }
+        const handler = handlers[type as RegisteredReducerType]
+        if (!handler) {
+          throw new Error(`Unsupported reducer type: ${String(type)}`)
+        }
+        const reducerDetails: ReducerDetails = {
+          reducerName,
+          type: getType(name, reducerName),
+        }
+        handler(reducerDetails, reducerDefinition as any, contextMethods)
       }
-      if (isAsyncThunkSliceReducerDefinition<State>(reducerDefinition)) {
-        handleThunkCaseReducerDefinition(
-          reducerDetails,
-          reducerDefinition,
-          contextMethods,
-          cAT,
-        )
-      } else {
-        handleNormalReducerDefinition<State>(
-          reducerDetails,
-          reducerDefinition,
-          contextMethods,
-        )
+    } else {
+      for (const [reducerName, reducerDefinition] of Object.entries(
+        options.reducers as SliceCaseReducers<State>,
+      )) {
+        const reducerDetails: ReducerDetails = {
+          reducerName,
+          type: getType(name, reducerName),
+        }
+        if ('reducer' in reducerDefinition) {
```

Comment:
:hand: silly byte-shaving - can we do `const handler = 'reducer' in reducerDefinition ? preparedReducerCreator : reducerCreator`, and then unconditionally do `handler(reducerDetails, reducerDefinition, contextMethods)` ?

Reply:
ya - originally this made typescript happier but since we're `as any`ing anyway, might as well